app/recipe/serializers.py

```diff
@@ -0,0 +1,79 @@
+from rest_framework import serializers
+from recipe.models import Ingredient, Recipe
+from django.utils.translation import gettext as _
+
+
+class IngredientSerializer(serializers.ModelSerializer):
+    """ Ingredient Serializer """
+
+    class Meta:
+        model = Ingredient
+        fields = ("name",)
+        read_only_fields = ("id",)
+
+
+class RecipeSerializer(serializers.ModelSerializer):
+    """ Recipe Serializer """
+
+    ingredients = IngredientSerializer(
+        many=True,
+    )
+
+    class Meta:
+        model = Recipe
+        fields = ("id", "name", "ingredients")
+        read_only_fields = ("id",)
+
+    def create(self, validated_data):
+        """When creating recipes, also create the ingredients
+        allowing for nested structures in the API"""
+        ingredients = validated_data.pop("ingredients", [])
+
+        # Raise if duplicate ingredients are listed
+        ingredient_names = [ingredient["name"] for ingredient in ingredients]
+        if len(ingredient_names) != len(set(ingredient_names)):
+            raise serializers.ValidationError(
+                _("A recipe cannot contain duplicate ingredients")
+            )
+
+        # Create the recipe, then the ingredients
+        instance = Recipe.objects.create(**validated_data)
+        for ingredient in ingredients:
+            Ingredient.objects.create(recipe=instance, **ingredient)
+        return instance
+
+    def update(self, instance, validated_data):
+        """
+        When updating recipes, we must also update ingredients:
+            Delete missing ingredients
+            Add new ingredients
+            Leaving matching ingredients as they are
+        """
+
+        ingredients = validated_data.pop("ingredients", [])
+        instance_ingredients = instance.ingredients.all()
+
+        # Use set operations to separate ingredients to create / delete
+        existing_ingredients = {
+            ingredient.name for ingredient in instance_ingredients
+        }
```

Comment:
Since `instance_ingredients` is QuerySet that will be executed here, it could be better to just ask for the values you need and not the "full" object (doesn't matter in this case, but in other cases with big objects it does).
Therefore, better do
```suggestion
        existing_ingredients = instance_ingredients.values_list("name", flat=True).distinct()
```

Reply:
didn't know about values_list, thanks  :+1: